llvm/lib/Transforms/InstCombine/InstCombineMulDivRem.cpp

```diff
@@ -1701,6 +1701,31 @@ static Instruction *foldFDivPowDivisor(BinaryOperator &I,
   return BinaryOperator::CreateFMulFMF(Op0, Pow, &I);
 }
 
+/// Convert div to mul if we have an sqrt divisor iff sqrt's operand is a fdiv
+/// instruction.
+static Instruction *foldFDivSqrtDivisor(BinaryOperator &I,
+                                        InstCombiner::BuilderTy &Builder) {
+  // X / sqrt(Y / Z) -->  X * sqrt(Z / Y)
+  Value *Op0 = I.getOperand(0), *Op1 = I.getOperand(1);
+  auto *II = dyn_cast<IntrinsicInst>(Op1);
+  if (!II || II->getIntrinsicID() != Intrinsic::sqrt || !II->hasOneUse() ||
```

Comment:
Would be cleaner to use PatternMatch. Something like

if (match(I, m_Sqrt(m_OneUse(m_FDiv(m_Value(Op0), m_Value(Op1))) {

}

Reply:
I agree but then we won't be able to check if `sqrt` has one use or `fdiv` has the necessary flags